src/rpc/rpc_payment_signature.cpp

```diff
@@ -0,0 +1,107 @@
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include "include_base_utils.h"
+#include "string_tools.h"
+#include "rpc_payment_signature.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "daemon.rpc.payment"
+
+#define TIMESTAMP_LEEWAY (60 * 1000000) /* 60 seconds, in microseconds */
+
+namespace cryptonote
+{
+  const boost::posix_time::ptime rpc_payment_epoch(boost::gregorian::date(1970, 1, 1));
+
+  std::string make_rpc_payment_signature(const crypto::secret_key &skey)
+  {
+    std::string s;
+    crypto::public_key pkey;
+    crypto::secret_key_to_public_key(skey, pkey);
+    crypto::signature sig;
+    const uint64_t now = (boost::posix_time::microsec_clock::universal_time() - rpc_payment_epoch).total_microseconds();
+    char ts[17];
+    snprintf(ts, sizeof(ts), "%16.16" PRIx64, now);
+    ts[16] = 0;
+    CHECK_AND_ASSERT_MES(strlen(ts) == 16, "", "Invalid time conversion");
+    crypto::hash hash;
+    crypto::cn_fast_hash(ts, 16, hash);
+    crypto::generate_signature(hash, pkey, skey, sig);
+    s = epee::string_tools::pod_to_hex(pkey) + ts + epee::string_tools::pod_to_hex(sig);
+    return s;
+  }
+
+  bool verify_rpc_payment_signature(const std::string &message, crypto::public_key &pkey, uint64_t &ts)
+  {
+    if (message.size() != 2 * sizeof(crypto::public_key) + 16 + 2 * sizeof(crypto::signature))
+    {
+      MDEBUG("Bad message size: " << message.size());
+      return false;
+    }
+    const std::string pkey_string = message.substr(0, 2 * sizeof(crypto::public_key));
+    const std::string ts_string = message.substr(2 * sizeof(crypto::public_key), 16);
+    const std::string signature_string = message.substr(2 * sizeof(crypto::public_key) + 16);
+    if (!epee::string_tools::hex_to_pod(pkey_string, pkey))
+    {
+      MDEBUG("Bad client id");
+      return false;
+    }
+    crypto::signature signature;
+    if (!epee::string_tools::hex_to_pod(signature_string, signature))
+    {
+      MDEBUG("Bad signature");
+      return false;
+    }
+    crypto::hash hash;
+    crypto::cn_fast_hash(ts_string.data(), 16, hash);
```

Comment:
After "sleeping on it", I think the best action is to require SSL or ZMQCurve for the pay-for-RPC feature. Both SSL and ZMQCurve provide data integrity, replay protection and privacy. The replay protection in both includes a server nonce such that the encrypted stream after the payload cannot be reused. SSL includes the timestamp in the nonce to protect against poor RNGs. This custom design allows replays for a short time period, provides no data integrity, and requires client and server clocks to be closely in sync (relative to each other).

The other question is how the client authentication is performed. SSL and ZMQCurve have client authentication builtin. Every callback could be forwarded the connection context (or just the client `optional<public_key>`). Given this current patch, a custom auth scheme is likely quicker to code (its already done!) but means maintaining some auth thing inside every RPC call.

Reply:
One other thing - in ZMQCurve the client pubkey is **not** sent in plaintext, but in SSL the client pubkey **is** sent in plaintext unless the server does a renegotiation on an existing connection OR TLS 1.3+ is used. So boo to SSL auth actually.

Reply:
I think I like that idea for signing the entire request. No replay AFAIK (I assume you did not account for the timestamp check (unless for a RPC that disables it on purpose), if there's still replay then I don't know about it).
I'm totally unclear on how to piggy back to SSL or Curvezmq for this though.

Reply:
I don't get a widget to reply to the comment below, so I'll comment here:

> The problem with the timestamp system is the relative differences between client and server
> can cause a rejection. Also the LEEWAY value means that replays are possible for a short-
> time frame. FWIW, Wireguard tracks timestamps per identity, and doesn't let them "rollback".

Do you think rejecting badly off times is really a problem ?

Leeway is not meant to prevent replay. The code that prevents replay is:
```
    if (ts < info.last_request_timestamp || (ts == info.last_request_timestamp && !same_ts))
    {
      MDEBUG("Invalid ts: " << ts << " <= " << info.last_request_timestamp);
      return false;
    }
```

Reply:
> I think I like that idea for signing the entire request. No replay AFAIK (I assume you did not account for the timestamp check (unless for a RPC that disables it on purpose), if there's still replay then I don't know about it).

Requiring SSL / ZMQ is likely less work, but signing the entire request is interesting too since it works even if someone botches the encrypted channel setup. I don't think its strictly necessary, but ¯\_(ツ)_/¯.

>  I'm totally unclear on how to piggy back to SSL or Curvezmq for this though.

Once authentication occurs within the encrypted+mac'ed stream, it never needs to be done through the remainder of the connection. Both systems provide encryption, integrity, and replay protection.

The simplest method is sending a random 32-bit value (as a JSON field) as an authentication token. The only downside is the server now has all of the authentication tokens unencrypted on the filesystem. The mitigation - a public key signature (as a JSON field) - can be to any arbitrary value (which is strange, for sure) since the contents of the stream should be unviewable and unreplayable.

There's also a new draft RFC for signing parts of a HTTP header with a public key (a new variant of HTTP auth), and existing HTTP auth. The latter would accept any user/pass value - it just needs to be consistent. Either might be awkward in this context but workable.

Using the builtin client authentication for SSL is easiest with Openssl 1.1.1+ since it can convert raw ed25519 to SSL certificates. Otherwise the authentication token has to be some SSL certificate generated by the client. ZMQ uses ed25519/curve25519 with libsodium. We already have code for SSL client authentication, but it would need to be adapted to accept any client certificate and then store that certificate in the connection context. ZMQCurve hasn't been implemented at all, but the process is similar.

I pointed out the ZMQ case to consider the possibility of making it work with both systems. Openssl 1.1.1+ also supports SSL ed25519 certificates -> raw value, so its definitely easier to use the same public key for SSL and ZMQ in that case.

> Do you think rejecting badly off times is really a problem ?

There is a decent probability that someone may run into a rejection due to an off-clock. I think this can be mostly mitigated by a nice error message.

The tricky case is with a "fast clock". The code allows any future timestamp, therefore someone that connected and then corrected their clock is now locked out or else needs to create a new account. A rare edge case, but an annoying one should it occur.

Reply:
On balance, I prefer the leeway system.